core/domain/exp_domain_test.py

```diff
@@ -435,6 +435,350 @@ def test_validate_exp_id(self):
+            *distinct_states)
+
+        self.assertEqual(len(spans), 3)
+        self.assertIs(spans[0].get_version(1), distinct_states[0])
+        self.assertIs(spans[1].get_version(2), distinct_states[1])
+        self.assertIs(spans[2].get_version(3), distinct_states[2])
+        self.assertEqual(spans[0].get_version_name(1), 'A')
+        self.assertEqual(spans[1].get_version_name(2), 'A')
+        self.assertEqual(spans[2].get_version_name(3), 'A')
+
+    def test_get_multi_versions_with_newly_constructed_span(self):
+        state = self.new_state_domain_obj()
+
+        span = exp_domain.StateVersionSpan(1, 'A', state)
+
+        self.assertEqual(span.get_multi_versions(1, 2), [state])
+        self.assertEqual(span.get_multi_version_names(1, 2), ['A'])
+
+    def test_get_multi_versions_with_extended_span(self):
+        distinct_states = [
+            self.new_state_domain_obj() for _ in python_utils.RANGE(3)]
+
+        (span,) = self.get_state_version_spans('A', None, *distinct_states)
+
+        self.assertEqual(span.get_multi_versions(1, 4), distinct_states)
+        self.assertEqual(span.get_multi_version_names(1, 4), ['A', 'A', 'A'])
+
+    def test_get_multi_versions_with_multi_spans(self):
+        distinct_states = [
+            self.new_state_domain_obj(interaction_id='MathEquationInput'),
+            self.new_state_domain_obj(interaction_id='TextInput'),
+            self.new_state_domain_obj(interaction_id='ItemSelectionInput'),
+        ]
+
+        spans = self.get_state_version_spans(
+            'A', lambda s1, s2: s1.interaction.id != s2.interaction.id,
+            *distinct_states)
+
+        self.assertEqual(len(spans), 3)
+        self.assertEqual(
+            spans[0].get_multi_versions(1, 2), distinct_states[0:1])
+        self.assertEqual(
+            spans[1].get_multi_versions(2, 3), distinct_states[1:2])
+        self.assertEqual(
+            spans[2].get_multi_versions(3, 4), distinct_states[2:3])
+        self.assertEqual(spans[0].get_multi_version_names(1, 2), ['A'])
+        self.assertEqual(spans[1].get_multi_version_names(2, 3), ['A'])
+        self.assertEqual(spans[2].get_multi_version_names(3, 4), ['A'])
+
+    def test_get_multi_versions_accepts_pythonic_ranges(self):
+        distinct_states = [
+            self.new_state_domain_obj() for _ in python_utils.RANGE(3)]
+
+        (span,) = self.get_state_version_spans(
+            'A', lambda s1, s2: s1.interaction.id != s2.interaction.id,
+            *distinct_states)
+
+        self.assertEqual(span.get_multi_versions(0, 20), distinct_states[:19])
+        self.assertEqual(span.get_multi_versions(2, 7), distinct_states[1:6])
+        self.assertEqual(span.get_multi_versions(6, 7), distinct_states[5:6])
+        self.assertEqual(span.get_multi_version_names(0, 20), ['A', 'A', 'A'])
+        self.assertEqual(span.get_multi_version_names(2, 7), ['A', 'A'])
+        self.assertEqual(span.get_multi_version_names(6, 7), [])
```

Comment:
@seanlip here's the motivation for the `get_multi_*` signature. WDYT?

Reply:
I think it's better to throw an error if the range is out of bounds. That way, there is also a guarantee about how many items will be in the returned list.

When I see ['A', 'A'], I don't know if it's 2 to 3, 3 to 4, etc.

Reply:
Done.